CraftingRevisions/Patches/GameManager_Awake.cs

```diff
@@ -4,15 +4,41 @@
 
 namespace CraftingRevisions.Patches
 {
-	[HarmonyPatch(typeof(GameManager), "Awake")]
-	internal class GameManager_Awake
+	// load the blueprints when the crafting panel is initialized
+	// (needed to be this late otherwise we have no instance of BlueprintManager)
+	[HarmonyPatch(typeof(Panel_Crafting), nameof(Panel_Crafting.Initialize))]
+	internal class Panel_Crafting_Initialize
 	{
 		private static void Postfix()
 		{
-			try { BlueprintMapper.MapBlueprints(); }
-			catch (System.Exception e)
+			// call TLD LoadAllUserBlueprints
+			InterfaceManager.m_Instance.m_BlueprintManager.LoadAllUserBlueprints();
+		}
+	}
+
+
+	// patched into postfix BlueprintManager.LoadAllUserBlueprints
+	// (BlueprintManager.RemoveUserBlueprints is not guaranteed to get called)
+	[HarmonyPatch(typeof(Il2CppTLD.Gear.BlueprintManager), nameof(Il2CppTLD.Gear.BlueprintManager.LoadAllUserBlueprints))]
+	internal class BlueprintManager_LoadAllUserBlueprints_Postfix
+	{
+		private static void Postfix(Il2CppTLD.Gear.BlueprintManager __instance)
+		{
+			// check we have a HasSet and it has contents
+			if (BlueprintManager.jsonUserBlueprints != null && BlueprintManager.jsonUserBlueprints.Count > 0)
 			{
-				MelonLogger.Error($"Blueprint Mapping failed: {e}");
+				// loop over the items
+				foreach (string jsonUserBlueprint in BlueprintManager.jsonUserBlueprints)
+				{
+					// load the blueprint into the game
+					bool loaded = __instance.LoadUserBlueprint(jsonUserBlueprint);
+
+#warning TODO - there appears to be no way to define an icon for this blueprint (UserBlueprintData still a WIP?)
+#warning TODO - get CraftingAudio working, is a string but unclear what value it wants. (see UserBlueprintData.MakeRuntimeWwiseEvent)
```

Comment:
I don't understand why this is an issue. Are the old strings not working?

Reply:
I've tried to debug this quite a bit, when the UserBlueprintData is loaded it calls MakeRuntimeWwiseEvent with the string but it throws an exception.

```
 During invoking native->managed trampoline
[22:43:48.206] Il2CppInterop.Runtime.Il2CppException: System.NullReferenceException: Object reference not set to an instance of an object.
--- BEGIN IL2CPP STACK TRACE ---
System.NullReferenceException: Object reference not set to an instance of an object.
  at TLD.Gear.UserBlueprintData.MakeRuntimeWwiseEvent (System.String eventName) [0x00000] in <00000000000000000000000000000000>:0
--- END IL2CPP STACK TRACE ---

   at Il2CppInterop.Runtime.Il2CppException.RaiseExceptionIfNecessary(IntPtr returnedException) in /home/runner/work/Il2CppInterop/Il2CppInterop/Il2CppInterop.Runtime/Il2CppException.cs:line 36
   at DMD<Il2CppTLD.Gear.UserBlueprintData::MakeRuntimeWwiseEvent>(String eventName)
   at (il2cpp -> managed) MakeRuntimeWwiseEvent(IntPtr , Il2CppMethodInfo* )
```

```
[Cpp2ILInjected.CallAnalysis.Calls(Type = typeof(AkSoundEngine), Member = "GetIDFromString")]
```
** Somewhere after here as I can patch the postfix of GetIDFromString and get the result >>
[22:43:48.203] AkSoundEngine_GetIDFromString_POST | PLAY_CRAFTINGCLOTH
[22:43:48.203] AkSoundEngine_GetIDFromString_POST | 1661985380
```
[Cpp2ILInjected.CallAnalysis.Calls(Type = typeof(ScriptableObject), Member = "CreateInstance")]
[Cpp2ILInjected.CallAnalysis.Calls(Type = typeof(Type), Member = "GetTypeFromHandle")]
[Cpp2ILInjected.CallAnalysis.Calls(Type = typeof(FieldInfo), Member = "SetValue")]
[Cpp2ILInjected.CallAnalysis.CallsDeduplicatedMethods(Count = 3)]
```

I'll continue to dig...

Reply:
> calls MakeRuntimeWwiseEvent with the string but it throws an exception.

I think this is caused by:
```json
"CraftingAudio":null,
```
Where the proper value would be:
```json
"CraftingAudio":"",
```
Because `MakeRuntimeWwiseEvent` might not support null input values.

> I'll continue to dig...

https://melonwiki.xyz/#/modders/debugging

Disclaimer: I've never done this for modding, but it's very useful for normal software development.